### ISSUE TITLE
Can't create migrations-record table

### ISSUE THREAD

**Issue by DonPangPang** [oracle/dotnet-db-samples#280]
Oracle Database Version: 18c
EF Core Version: 7.0.3
Oracle.EntityFramewrokCore Version:7.21.9
Oracle.ManagedDataAccess.Core Version:3.21.90
Platform: windows 11

When i use `dbContext.Database.Migrete()` to update datebase. The first time, it will create a table named ”__EFMigrationsHistory“, but the sql is wrong.
![image](https://user-images.githubusercontent.com/49866401/227450468-9a3abd11-a6d1-4590-9abb-57e8b16f1af7.png)

**Comment by alexkeh** [comment]
@DonPangPang Can you turn on ODP.NET tracing and share the trace file? I can then take a look at what exactly is wrong with the CREATE statement. You can use something like the following to turn on tracing. Insert before you open your first ODP.NET connection.

```
Oracle.ManagedDataAccess.Client.OracleConfiguration.TraceFileLocation = @"C:\traces";
Oracle.ManagedDataAccess.Client.OracleConfiguration.TraceLevel = 7;
```

**Comment by DonPangPang** [comment]
Sorry, I just saw your reply, let me describe the problem.

This project uses the Database-First model, but later introduces efcore's Migrations feature. When the first migration occurs, efcore creates a table named `__MigrationsHistory`.  The SQL statement generated by efcore is `Create table __MigrationsHistory(...)`, but efcore only supports the format as
```
BEGIN
    EXECUTE IMMEDIATE 'Create table ...';
END;
```

This issue does not affect the work of ef-tools, but it does affect the use of `DbContext.Database.Migrate()` in projects.

**Comment by alexkeh** [comment]
While this may be a possible use case, using scaffolding followed by a migration with the same migration history table isn't an explicitly supported scenario. Are you aware of any MS provider that supports this usage scenario?

MS doc says that [you can scaffold once, then switch to migrations](https://learn.microsoft.com/en-us/ef/core/managing-schemas/scaffolding/?tabs=dotnet-core-cli#scaffold-once-only). It doesn't say whether the migration history table can be used in that scenario.

Nonetheless, one workaround is to change the name of the migrations history table so that a different one is created for migrations and doesn't conflict with the scaffolded history table. You can generate migrations history table, such as using:

`optionsBuilder.UseOracle(@"User Id=MYSCHEMA;Password=PASSWORD;Data Source=DATASOURCE", x => x.MigrationsHistoryTable("__CustomMigrationsHistory", "MYSCHEMA"));`

**Comment by DonPangPang** [comment]
Thanks for the reply as well as the solution.

I solved this problem by rewriting part of the SQL generator. But I think this is still a problem, in other database support programs, it is only necessary to call the `DbContext.Database.Migrate()` method to generate the `__MigrationsHistory` table.

**Comment by alexkeh** [comment]
I'm closing the issue since you found a solution.

In terms of supporting this scenario, which other providers support this scenario in their `DbContext.Database.Migrate()` method you're aware of?

**Comment by alexkeh** [comment]
I filed an enhancement request (ER) 35256973 to track this request.

What changes did you make to the SQL generator?

**Comment by DonPangPang** [comment]
I did the check at startup, manually checked the existence of the `__MigrationsHistory` table before `DbContext.Database.Migrate()` with ADO.NET statements, and then created it using
```
BEGIN
    EXECUTE IMMEDIATE 'Create table __MigrationsHistory(...)';
END;
```

For changes to other table structures, I rewrote `MigrationsSqlGenerator`.
![image](https://user-images.githubusercontent.com/49866401/230127616-31148483-4e59-4e3a-b84c-7b64fba7899b.png)
This `SqlManager.Block` method is to convert SQL statements to 
```
BEGIN
    EXECUTE IMMEDIATE 'sql...';
END;
```
format.

**Comment by alexkeh** [comment]
@DonPangPang 
In investigating this issue, the Oracle .NET team is not able to replicate this issue. We tried to replicate the issue with following scenarios.

Scenario 1:
 - Scaffolded a table and applied first migration using EF tools
 - Again scaffolded same table to different context and run the app using "dbContext.Database.Migrate()" method to apply the migration to same database.
 - with each new migration applied our provider generates below PL/SQL command which is not throwing any error when migration is applied using "dbContext.Database.Migrete()"

```
BEGIN
      EXECUTE IMMEDIATE 'CREATE TABLE
      "__EFMigrationsHistory" (
          "MigrationId" NVARCHAR2(150) NOT NULL,
          "ProductVersion" NVARCHAR2(32) NOT NULL,
          CONSTRAINT "PK___EFMigrationsHistory" PRIMARY KEY ("MigrationId")
      )';
      END;
```

Scenario 2:
Repeated same above steps with following change but not getting any error:

```
optionsBuilder.UseOracle(@"User Id=MYSCHEMA;Password=PASSWORD;Data
Source=DATASOURCE"
, x => x.MigrationsHistoryTable("__CustomMigrationsHistory", "MYSCHEMA"));

```
In order to understand this issue better, Can you provide more detailed replication steps which can help generate the same PL/SQL command (without "EXECUTE IMMEDIATE" keywords)?

**Comment by DonPangPang** [comment]
> @DonPangPang In investigating this issue, the Oracle .NET team is not able to replicate this issue. We tried to replicate the issue with following scenarios.
> 
> Scenario 1:
> 
> * Scaffolded a table and applied first migration using EF tools
> * Again scaffolded same table to different context and run the app using "dbContext.Database.Migrate()" method to apply the migration to same database.
> * with each new migration applied our provider generates below PL/SQL command which is not throwing any error when migration is applied using "dbContext.Database.Migrete()"
> 
> ```
> BEGIN
>       EXECUTE IMMEDIATE 'CREATE TABLE
>       "__EFMigrationsHistory" (
>           "MigrationId" NVARCHAR2(150) NOT NULL,
>           "ProductVersion" NVARCHAR2(32) NOT NULL,
>           CONSTRAINT "PK___EFMigrationsHistory" PRIMARY KEY ("MigrationId")
>       )';
>       END;
> ```
> 
> Scenario 2: Repeated same above steps with following change but not getting any error:
> 
> ```
> optionsBuilder.UseOracle(@"User Id=MYSCHEMA;Password=PASSWORD;Data
> Source=DATASOURCE"
> , x => x.MigrationsHistoryTable("__CustomMigrationsHistory", "MYSCHEMA"));
> ```
> 
> In order to understand this issue better, Can you provide more detailed replication steps which can help generate the same PL/SQL command (without "EXECUTE IMMEDIATE" keywords)?

The first step is not `Scaffolded a table and applied first migration using EF tools`, should be `Database-first`. Then, create migrations and use `dbContext.Database.Migrate()`

**Comment by alexkeh** [comment]
@DonPangPang 
We're still not sure what specific steps you use. We're looking for how to create a reproducible test case. A few questions:
1.	While applying 'database-first (scaffolding)', do you already have "__EFMigrationsHistory" table created on the database?
2.	Can you list the order of EF tool commands used before calling "dbContext.Database.Migrate()" method?
3.	What is the model he is trying to migrate? Can you provide all .NET classes necessary to reproduce the issue for you call the "dbContext.Database.Migrate()" method?

**Comment by DonPangPang** [comment]
> @DonPangPang We're still not sure what specific steps you use. We're looking for how to create a reproducible test case. A few questions:
> 
> 1. While applying 'database-first (scaffolding)', do you already have "__EFMigrationsHistory" table created on the database?
> 2. Can you list the order of EF tool commands used before calling "dbContext.Database.Migrate()" method?
> 3. What is the model he is trying to migrate? Can you provide all .NET classes necessary to reproduce the issue for you call the "dbContext.Database.Migrate()" method?

1. No "__EFMigrationsHistory" table
2. `db-fisrt`, and use once `dotnet-ef migrations add init` to get `Snapshot`, then drop the `init migration`, because the db already exists.
3. The migration model can be any change. You can make as many Entity-Changes as you like.

Then the provider will be try to create the `__EFMigrationsHistory` table fail.

**Comment by alexkeh** [comment]
We're still not the error. Here are the steps we undertook:

1.	Created new schema and one table. (the "__EFMigrationsHistory" table is not present)
2.	Scaffolded(db-first) the table using below command.
a.	This step has generated one DbContext class named "Context1.cs"
b.	and one entity class named "Tbl.cs"
`Scaffold-DbContext "User Id=<USER>;Password=<PASSWORD>;Data Source= tns descriptor" Oracle.EntityFrameworkCore -OutputDir scaffold1 -Force -Context Context1`
3.	Added migration named 'init' which has generated 'Migrations' folder with "Context1ModelSnapshot.cs" and "20230502121734_init.cs" files
a.	In this step, I am able to get the snapshot file.
4.	Drop the 'init' migration
a.	In this step, I have just deleted the "20230502121734_init.cs" file (If I use Remove-Migration command then it drops the whole 'Migrations' folder including snapshot file)
5.	Made Entity changes in "Table.cs" and "Context1.cs" files (changed cs and db datatypes)
6.	Added another migration named "second" which has generated below Up method in "20230502122034_init.cs" file. 
```
protected override void Up(MigrationBuilder migrationBuilder) 
        {
            migrationBuilder.AlterColumn<int>(
                name: "NUMBER1",
                schema: "TS_SCAFFOLDTESTCASE",
                table: "TBL",
                type: "NUMBER(10)",
                nullable: true,
                oldClrType: typeof(byte),
                oldType: "NUMBER(4)",
                oldPrecision: 4,
                oldNullable: true);
        }
```
7.	Called context.Database.Migrate(); method.
8.	Provider has successfully created the "__EFMigrationsHistory" table and applied the entity changes without any error.
